man/man8/tcm_nvinfo.8
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "tcm_nvinfo 8"
.TH tcm_nvinfo 8 "2018-07-12" "TCM Management" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
tpm_nvinfo
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBtpm_nvinfo [\s-1OPTIONS\s0]\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBtpm_nvinfo\fR displays information about defined \s-1NVRAM\s0 areas. The user may
restrict it to display only a single \s-1NVRAM\s0 area by providing an index or may
choose ton only list available \s-1NVRAM\s0 areas indices without displaying the
\&\s-1NVRAM\s0 areas' detailed information.
.PP
The following options are supported:
.IP "\fB\-h, \-\-help\fR" 4
.IX Item "-h, --help"
Display command usage info.
.IP "\fB\-v, \-\-version\fR" 4
.IX Item "-v, --version"
Display command version info.
.IP "\fB\-l, \-\-log [none|error|info|debug]\fR" 4
.IX Item "-l, --log [none|error|info|debug]"
Set logging level.
.IP "\fB\-i, \-\-index\fR (optional parameter)" 4
.IX Item "-i, --index (optional parameter)"
Only display the \s-1NVRAM\s0 area with the given index. If this option is not
provided all \s-1NVRAM\s0 areas will be displayed.
.IP "\fB\-n, \-\-list\-only\fR (optional parameter)" 4
.IX Item "-n, --list-only (optional parameter)"
Only display the defined \s-1NVRAM\s0 areas' indices.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBtpm_nvdefine\fR(8), \fBtpm_nvread\fR(8), \fBtpm_nvwrite\fR(8), \fBtpm_nvrelease\fR(8)
.SH "REPORTING BUGS"
.IX Header "REPORTING BUGS"
Report bugs to <trousers\-users@lists.sourceforge.net>


src/tpm_mgmt/tcm_nvinfo.c
/*
 * The Initial Developer of the Original Code is International
 * Business Machines Corporation. Portions created by IBM
 * Corporation are Copyright (C) 2005 International Business
 * Machines Corporation. All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the Common Public License as published by
 * IBM Corporation; either version 1 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Common Public License for more details.
 *
 * You should have received a copy of the Common Public License
 * along with this program; if not, a copy can be viewed at
 * http://www.opensource.org/licenses/cpl1.0.php.
 */

#include <limits.h>
#include <arpa/inet.h>

#include "tcm_tspi.h"
#include "tcm_utils.h"
#include "tcm_nvcommon.h"


static BOOL nvindex_set;
static unsigned int nvindex;
static BOOL list_only;
TSM_HCONTEXT hContext = 0;

static int parse(const int aOpt, const char *aArg)
{

	switch (aOpt) {
	case 'i':
		if (parseHexOrDecimal(aArg, &nvindex, 0, UINT_MAX,
				      "NVRAM index") != 0)
			return -1;

		nvindex_set = TRUE;
		list_only = FALSE;

		break;

	case 'n':
		list_only = TRUE;
		nvindex_set = FALSE;
		break;

	default:
		return -1;
	}
	return 0;
}


static void help(const char* aCmd)
{
	logCmdHelp(aCmd);
	logNVIndexCmdOption();
	logCmdOption("-n, --list-only",
		     _("Only list the defined NVRAM areas' indices."));
}


static void pcrInfoShortDisplay(TCM_PCR_INFO_SHORT *tpis, const char *type)
{
	UINT16 i, c;

	c = 0;

	logMsg("PCR %sselection:\n", type);

	for (i = 0; i < tpis->pcrSelection.sizeOfSelect * 8; i++) {
		if (tpis->pcrSelection.pcrSelect[(i / 8)] & (1 << (i & 0x7))) {
			if (!c)
				logMsg(" PCRs	 : ");
			if (c)
				logMsg(", ");
			printf("%d", i);
			c++;
		}
	}

	if (c)
		logMsg("\n");

	if (tpis->localityAtRelease) {
		if (tpis->localityAtRelease == 0x1f) {
			logMsg(" Localities   : ALL\n");
		} else {
			logMsg(" Localities   : 0x%01x\n", tpis->localityAtRelease);
		}
	}

	if (c) {
		logMsg(" Hash	 : ");
		for (i = 0; i < 32; i++)
			logMsg("%02x", tpis->digestAtRelease.digest[i]);
		logMsg("\n");
	}
}


static void nvindexDisplay(TSM_HTCM hTpm, UINT32 nvindex)
{
	TSM_RESULT res;
	char *buffer;
	TCM_NV_DATA_PUBLIC *nvpub = NULL;

	logMsg("NVRAM index   : 0x%08x (%u)\n", nvindex, nvindex);

	res = getNVDataPublic(hTpm, nvindex, &nvpub);

	if (res != TSM_SUCCESS)
		goto out;

	pcrInfoShortDisplay(&nvpub->pcrInfoRead , "read  ");
	pcrInfoShortDisplay(&nvpub->pcrInfoWrite, "write ");

	buffer = printValueAsStrings((unsigned int)nvpub->permission.attributes,
	                             permvalues);

	logMsg("Permissions   : 0x%08x (%s)\n", nvpub->permission.attributes, buffer);
	free(buffer);
	buffer = NULL;

	logMsg("bReadSTClear  : %s\n", nvpub->bReadSTClear ? "TRUE" : "FALSE");
	logMsg("bWriteSTClear : %s\n", nvpub->bWriteSTClear ? "TRUE" : "FALSE");
	logMsg("bWriteDefine  : %s\n", nvpub->bWriteDefine ? "TRUE" : "FALSE");

	logMsg("Size          : %d (0x%x)\n", nvpub->dataSize, nvpub->dataSize);


     out:
	freeNVDataPublic(nvpub);

	return;
}


int main(int argc, char **argv)
{
	TSM_HTCM hTpm;
	UINT32 ulResultLen;
	BYTE *pResult = NULL;
	int iRc = -1;
	unsigned int i;
	struct option hOpts[] = {
		{"index"    , required_argument, NULL, 'i'},
		{"list-only",       no_argument, NULL, 'n'},
		{NULL       ,       no_argument, NULL, 0},
	};

	initIntlSys();

	if (genericOptHandler
		    (argc, argv, "i:o:n", hOpts,
		     sizeof(hOpts) / sizeof(struct option), parse, help) != 0)
		goto out;

	if (contextCreate(&hContext) != TSM_SUCCESS)
		goto out;

	if (contextConnect(hContext) != TSM_SUCCESS)
		goto out_close;

	if (contextGetTpm(hContext, &hTpm) != TSM_SUCCESS)
		goto out_close;


	if (getCapability(hTpm, TSM_TCMCAP_NV_LIST, 0, NULL,
			  &ulResultLen, &pResult) != TSM_SUCCESS) {
		goto out_close;
	}

	if (list_only) {
		logMsg(_("The following NVRAM areas have been defined:\n"));
	}

	for (i = 0; i < ulResultLen/sizeof(UINT32); i++) {
		UINT32 nvi;
		nvi = Decode_UINT32(pResult + i * sizeof(UINT32));

		if (list_only) {
			logMsg("0x%08x (%d)\n", nvi, nvi);
		} else {
			if ((nvindex_set && nvi == (UINT32)nvindex) ||
			     !nvindex_set) {
				nvindexDisplay(hTpm, nvi);
				logMsg("\n");
			}
		}
	}

	iRc = 0;

      out_close:
	contextClose(hContext);

      out:

	return iRc;
}
